### ISSUE TITLE
Segfault in Ostap

### ISSUE THREAD

**Issue by bash-spbu** [JetBrains-Research/Lama#7]
### Problem:
I faced some strange behaviour when wanted to use Do-notation with Ostap's parsers.

Consider the following example:
```haskell
import Ostap;

fun pure(v) {
  empty @ lift(v)
}

local parse = 
  token(".") |> fun(e1) { 
  token(".") |> fun(e2) {
  pure(Dot(e1, e2)) }};

case parseString(parse |> bypass(eof), "..") of
   Succ(_)       -> printf("Yes")
 | Fail(_, _, _) -> printf("No")
esac
```
When I compile & run this I get a segfault somewhere (unfortunatelly GDB cannot locate it precisely).

But if we inline `pure` definition (what should not change the semantics, should it?), everything will be OK:
```haskell
import Ostap;

fun pure(v) {
  empty @ lift(v)
}

local parse = 
  token(".") |> fun(e1) { 
  token(".") |> fun(e2) {
  empty @ lift(Dot(e1, e2)) }};

case parseString(parse |> bypass(eof), "..") of
   Succ(_)       -> printf("Yes")
 | Fail(_, _, _) -> printf("No")
esac
```
Result: `Yes`

### Expected behaviour:
Same semantics for both snippets.

### Environment:
- lamac varsion: Version master, b7ec1c2e6, Mon Mar 23 00:51:00 2020 +0300

**Comment by dboulytchev** [comment]
Looking like some tail-call optimization issue.

**Comment by bash-spbu** [comment]
Also I probably should mention that this example is not an issue when there is only one token combinator:

```haskell
import Ostap;

fun pure(v) {
  empty @ lift(v)
}

local parse = 
  token(".") |> fun(e1) { 
  pure(Dot(e1)) };

case parseString(parse |> bypass(eof), ".") of
   Succ(_)       -> printf("Yes")
 | Fail(_, _, _) -> printf("No")
esac
```
Works as expected.

Maybe it will help to localize the problem.

**Comment by dboulytchev** [comment]
Fixed in 7748144.
